iceoryx_utils/include/iceoryx_utils/internal/concurrent/lockfree_queue/index_queue.inl

```diff
@@ -0,0 +1,255 @@
+namespace iox
+{
+namespace concurrent
+{
+template <uint64_t Capacity, typename ValueType>
+constexpr IndexQueue<Capacity, ValueType>::IndexQueue(ConstructEmpty_t) noexcept
+    : m_readPosition(Index(Capacity))
+    , m_writePosition(Index(Capacity))
+{
+}
+
+template <uint64_t Capacity, typename ValueType>
+IndexQueue<Capacity, ValueType>::IndexQueue(ConstructFull_t) noexcept
+    : m_readPosition(Index(0u))
+    , m_writePosition(Index(Capacity))
+{
+    for (uint64_t i = 0u; i < Capacity; ++i)
+    {
+        m_cells[i].store(Index(i));
+    }
+}
+
+template <uint64_t Capacity, typename ValueType>
+constexpr uint64_t IndexQueue<Capacity, ValueType>::capacity() const noexcept
+{
+    return Capacity;
+}
+
+template <uint64_t Capacity, typename ValueType>
+void IndexQueue<Capacity, ValueType>::push(const ValueType index) noexcept
+{
+    // we need the CAS loop here since we may fail due to concurrent push operations
+    // note that we are always able to succeed to publish since we have
+    // enough capacity for all unique indices used
+
+    // case analyis
+    // (1) loaded value is exactly one cycle behind:
+    //     value is from the last cycle
+    //     we can try to publish
+    // (2) loaded value has the same cycle:
+    //     some other push has published but not updated the write position
+    //     help updating the write position
+    // (3) loaded value is more than one cycle behind:
+    //     this should only happen due to wrap around when push is interrupted for a long time
+    //     reload write position and try again
+    //     note that a complete wraparound can lead to a false detection of 1) (ABA problem)
+    //     but this is very unlikely with e.g. a 64 bit value type
+    // (4) loaded value is some cycle ahead:
+    //     write position is outdated, there must have been other pushes concurrently
+    //     reload write position and try again
+
+    constexpr bool notPublished = true;
```

Comment:
according to our style guilde, constants should be UPPER_CAMEL_CASE

Reply:
Yes, but this just looks bad here...I would prefer a while true loop over that. Another commit just for this seems excessive.

Reply:
I would also be fine with UpperCamelCase

Reply:
Technically reserved for types...oh well.